### PR TITLE
Add initial OMIS creation journey

### PR DESCRIPTION
**Pull request by teneightfive**
This creates the basis for the OMIS order creation journey.

It currently supports the company name being passed in as part of the URL which then allows datasets like the contacts to be populated from it.

It support being able to go back to steps within the journey and update the fields.

There is currently no way to get to the journey by navigating pages so you will have to access using the URL: `/omis/create/:companyId`. There is a link to the journey on heroku in the _what it looks like_ section.

## Todo

- [x] Controller/middleware tests
- [x] Reset session after submission

## Not included in this change

- Field validation
- Order view (it will currently `404` after submission)
- Final content (includes placeholder copy)

## What it looks like

[Journey link](https://datahub-beta2-pr-355.herokuapp.com/omis/create/dcdabbc9-1781-e411-8955-e4115bead28a)

omis-create

### CODE REVIEW

File: assets/stylesheets/components/_summary-table.scss [hunk 1]
```diff
@@ -0,0 +1,52 @@
+@import "../settings/colours";
+@import "../settings/layout";
+@import "../tools/mixins/conditionals";
+@import "../tools/mixins/typography";
+
+.c-summary-table {
```
Comment (ztolley) [inline assets/stylesheets/components/_summary-table.scss]: How is this different from the existing table styles available

File: src/apps/omis/apps/create/controllers/base.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const Controller = require('hmpo-form-wizard').Controller
```
Comment (ztolley) [inline src/apps/omis/apps/create/controllers/base.js]: I think wee need top all discuss hmpo as this is completely different to how the other forms work and means the code is now inconsistent, also architectural changes like this should be agreed as a team, maybe after a spike and demo

File: src/apps/omis/apps/create/controllers/company-details.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+const { sortBy } = require('lodash')
```
Comment (ztolley) [inline src/apps/omis/apps/create/controllers/company-details.js]: The filename confused me and I thought it was adding a company

File: src/apps/omis/apps/create/controllers/market.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const Controller = require('./base')
+const metadataRepo = require('../../../../../lib/metadata')
+
+class MarketController extends Controller {
+  configure (req, res, next) {
+    req.form.options.fields.primary_market.options = metadataRepo.countryOptions.map((item) => {
+      return {
+        value: item.id,
```
Comment (ztolley) [inline src/apps/omis/apps/create/controllers/market.js]: Always annoys me we have to do this

File: src/apps/omis/apps/create/controllers/summary.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      values.ita = find(advisers.results, { id: values.ita })
+
+      next(err, values)
+    })
+  }
+
+  saveValues (req, res, next) {
+    const data = req.sessionModel.toJSON()
+
+    // clean unnedeed properties
```
Comment (ztolley) [inline src/apps/omis/apps/create/controllers/summary.js]: unnedeed ?

File: src/apps/omis/apps/create/controllers/summary.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    unset(data, 'csrf-secret')
+    unset(data, 'errors')
+
+    Order.save(req.session.token, data)
+      .catch(next)
+      .then((order) => {
+        console.log('RESETTING SESSION')
```
Comment (ztolley) [inline src/apps/omis/apps/create/controllers/summary.js]: don't we have a logger?

File: src/apps/omis/apps/create/index.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+const router = require('./router')
+
+module.exports = {
+  displayName: 'Create new order',
+  mountpath: '/create',
```
Comment (ztolley) [inline src/apps/omis/apps/create/index.js]: mountPath?

File: src/apps/omis/middleware/params.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+const logger = require('../../../../config/logger')
+const { getInflatedDitCompany } = require('../../companies/services/data')
+
+async function getCompany (req, res, next, companyId) {
+  try {
+    res.locals.company = await getInflatedDitCompany(req.session.token, companyId)
```
Comment (ztolley) [inline src/apps/omis/middleware/params.js]: Do you REALLY need the inflated version? We actually need to phase this method out as it's no longer needed and when you need details for contacts etc this should be a separate call.

File: src/apps/omis/middleware/params.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    next()
+  } catch (e) {
+    logger.error(e)
+    res.redirect('/')
```
Comment (ztolley) [inline src/apps/omis/middleware/params.js]: not next(e)?

File: src/apps/omis/views/create/start.njk [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{% extends "_layouts/form-wizard-step.njk" %}
+
+{% block form %}
+  {% call MultiStepForm({
+    buttonText: 'Start'
+  }) %}
+    <p>You will need to be aware of these things before you start this process:</p>
+
+    <ul class="list-disc">
```
Comment (ztolley) [inline src/apps/omis/views/create/start.njk]: namespace the class?

**Comment by teneightfive** [comment]
>We all need to discuss HMPO as a team

Agreed but it is moving the direction of building forms dynamically that we have discussed. It would be more efficient that manually creating fields in views.

>There is no way of knowing where you are in the process and knowing what back takes you to

This is out of scope for this change and we have a ticket to track the problem of journey navigation. There is also some [research](https://www.gov.uk/service-manual/design/progress-indicators) done around GOV.UK journeys that [progress indicators](https://paper.dropbox.com/doc/Progress-indicators-FKi7F1KBjks0jGG5vMdVg) don't actually add much value so maybe we start with it off and see if it causes users problems.

File: assets/stylesheets/components/_summary-table.scss (hunk 1 above)
Comment (teneightfive) [inline assets/stylesheets/components/_summary-table.scss]: It specifically adjusts the layout to be stacked for smaller devices which the current table styling doesn't cater for.

It also applies style to allow the controls to be consistently placed and aligned.

File: src/apps/omis/apps/create/controllers/base.js (hunk 2 above)
Comment (teneightfive) [inline src/apps/omis/apps/create/controllers/base.js]: Yes but this is also an interim replacement and being developed and maintained by a separate team so we shouldn't be restricted completely. 

It also moves to a form builder architecture which will make maintaining journeys within this app more straightforward.

File: src/apps/omis/apps/create/controllers/company-details.js (hunk 3 above)
Comment (teneightfive) [inline src/apps/omis/apps/create/controllers/company-details.js]: Not sure what would be a better name? It's a controller for adding company details as part of this journey.

File: src/apps/omis/apps/create/controllers/market.js (hunk 4 above)
Comment (teneightfive) [inline src/apps/omis/apps/create/controllers/market.js]: Completely agreed. Would like to find a better way to format these.

File: src/apps/omis/apps/create/controllers/summary.js (hunk 6 above)
Comment (teneightfive) [inline src/apps/omis/apps/create/controllers/summary.js]: Ah. Good spot. This needs addressing. And needs to actually reset the session.

File: src/apps/omis/apps/create/index.js (hunk 7 above)
Comment (teneightfive) [inline src/apps/omis/apps/create/index.js]: This matches the [naming convention](http://expressjs.com/en/api.html#app.mountpath) used by Express and also used in our other sub-apps.

File: src/apps/omis/middleware/params.js (hunk 8 above)
Comment (teneightfive) [inline src/apps/omis/middleware/params.js]: Fair point. I was using the methods that have previously been used. Will see if there is an alternative.

File: src/apps/omis/middleware/params.js (hunk 9 above)
Comment (teneightfive) [inline src/apps/omis/middleware/params.js]: We may need to specifically handle this but I didn't want to present the user with a `404` from the API when this catch happens.

This will be addressed as we explore other entry points into this journey.

File: src/apps/omis/views/create/start.njk (hunk 10 above)
Comment (teneightfive) [inline src/apps/omis/views/create/start.njk]: This is using the pattern we currently adopt and personally I'd rather we don't have to apply a class to re-apply the default list style to a list.

But when I applied that style it had a lot of knock-on affects. We should really have a typography and style review sooner rather than later.

File: src/apps/omis/apps/create/controllers/market.js (hunk 4 above)
Comment (tyom) [inline src/apps/omis/apps/create/controllers/market.js]: Maybe it can be moved into a helper function somewhere in the lib?

File: assets/stylesheets/components/_summary-table.scss [hunk 11]
```diff
@@ -0,0 +1,53 @@
+@import "../settings/colours";
```
Comment (tyom) [inline assets/stylesheets/components/_summary-table.scss]: Do we need this granularity? You could just import `../settings` and `../tools` or even just `../tools` which imports settings.

File: src/apps/omis/apps/create/controllers/base.js (hunk 2 above)
Comment (tyom) [inline src/apps/omis/apps/create/controllers/base.js]: `const { Controller } = require('hmpo-form-wizard')`?

File: src/apps/omis/apps/create/controllers/assign-ita.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
+const { sortBy } = require('lodash')
+
+const Controller = require('./base')
+const { getAdvisers } = require('../../../../adviser/repos')
+
+class AssignItaController extends Controller {
+  async configure (req, res, next) {
+    const advisers = await getAdvisers(req.session.token)
+    const options = advisers.results.map((adviser) => {
```
Comment (tyom) [inline src/apps/omis/apps/create/controllers/assign-ita.js]: I'd suggest extracting this function into a standalone one. Maybe in transformers?

File: src/apps/omis/apps/create/controllers/company-details.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+class CompanyDetailsController extends Controller {
+  configure (req, res, next) {
+    const company = res.locals.company
+    let contacts = []
+
+    if (company) {
+      contacts = company.contacts.map((option) => {
+        return {
+          value: option.id,
```
Comment (tyom) [inline src/apps/omis/apps/create/controllers/company-details.js]: This is popping up everywhere. I reckon we should move it into a global transformer. In a separate PR I guess.

File: src/apps/omis/middleware/params.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+const logger = require('../../../../config/logger')
+const { getInflatedDitCompany } = require('../../companies/services/data')
+
+async function getCompany (req, res, next, companyId) {
```
Comment (tyom) [inline src/apps/omis/middleware/params.js]: Should this go in models?

File: src/apps/omis/apps/create/controllers/company-details.js (hunk 13 above)
Comment (teneightfive) [inline src/apps/omis/apps/create/controllers/company-details.js]: I have extracted out to an OMIS level transformer but maybe feels like it should live higher level and could be used in lots of other places. Maybe the changing existing behaviour could be a separate PR but will add it with this one now.

File: src/apps/omis/middleware/params.js (hunk 14 above)
Comment (teneightfive) [inline src/apps/omis/middleware/params.js]: Maybe if we created a model within `companies` but I suppose we haven't firmed up that decision yet, right?

Comment (teneightfive) [inline src/apps/omis/middleware/params.js]: These param level functions that basically set things on `res.locals` feel like they could start living higher too and being used in multiple apps to avoid repeating setting these values.

Definitely different change though.

File: src/apps/omis/transformers.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
+const { isArray } = require('lodash')
+
+function transformToOptions (items, labelIteratee = (item) => item.name) {
```
Comment (tyom) [inline src/apps/omis/transformers.js]: Should we move it level up to global transformers to other apps can use it?

File: src/apps/omis/transformers.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  if (!isArray(items)) { return [] }
```
Comment (tyom) [inline src/apps/omis/transformers.js]: Should it return empty array if it's not an array to begin with? It could potentially swallow a user error.

Comment (teneightfive) [inline src/apps/omis/transformers.js]: Fair point. Will remove this.

File: src/apps/omis/middleware/params.js (hunk 14 above)
Comment (tyom) [inline src/apps/omis/middleware/params.js]: Do we need this in `params` file? They seem rather small, could they live in one `middleware.js` file?

File: test/unit/apps/transformers.test.js [hunk 17]
```diff
@@ -1,4 +1,4 @@
-const transformers = require('~/src/apps/omis/transformers')
+const transformers = require('~/src/apps/transformers')
 
 describe('OMIS transformers', () => {
```
Comment (tyom) [inline test/unit/apps/transformers.test.js]: They are not OMIS anymore.

Comment (tyom) [inline test/unit/apps/transformers.test.js]: Global data transformers?

Comment (teneightfive) [inline test/unit/apps/transformers.test.js]: Damn. Good spot.
